diff --git a/rts/package.conf.in b/rts/package.conf.in
index 9bdbf365..993d43cc 100644
--- a/rts/package.conf.in
+++ b/rts/package.conf.in
@@ -179,7 +179,9 @@ ld-options:
          , "-Wl,-u,_hs_xchg8"
          , "-Wl,-u,_hs_xchg16"
          , "-Wl,-u,_hs_xchg32"
+#if WORD_SIZE_IN_BITS == 64
          , "-Wl,-u,_hs_xchg64"
+#endif
          , "-Wl,-u,_hs_atomicread8"
          , "-Wl,-u,_hs_atomicread16"
          , "-Wl,-u,_hs_atomicread32"
@@ -293,7 +295,9 @@ ld-options:
          , "-Wl,-u,hs_xchg8"
          , "-Wl,-u,hs_xchg16"
          , "-Wl,-u,hs_xchg32"
+#if WORD_SIZE_IN_BITS == 64
          , "-Wl,-u,hs_xchg64"
+#endif
          , "-Wl,-u,hs_atomicread8"
          , "-Wl,-u,hs_atomicread16"
          , "-Wl,-u,hs_atomicread32"
diff --git a/rts/rts.cabal.in b/rts/rts.cabal.in
index 0a06414d..2d884c1f 100644
--- a/rts/rts.cabal.in
+++ b/rts/rts.cabal.in
@@ -308,7 +308,6 @@ library
          "-Wl,-u,_hs_xchg8"
          "-Wl,-u,_hs_xchg16"
          "-Wl,-u,_hs_xchg32"
-         "-Wl,-u,_hs_xchg64"
          "-Wl,-u,_hs_atomicread8"
          "-Wl,-u,_hs_atomicread16"
          "-Wl,-u,_hs_atomicread32"
@@ -389,7 +388,6 @@ library
          "-Wl,-u,hs_xchg8"
          "-Wl,-u,hs_xchg16"
          "-Wl,-u,hs_xchg32"
-         "-Wl,-u,hs_xchg64"
          "-Wl,-u,hs_atomicread8"
          "-Wl,-u,hs_atomicread16"
          "-Wl,-u,hs_atomicread32"
